examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py

```diff
@@ -0,0 +1,598 @@
+        # section!
+        return original_output
+
+
+"""
+### Inject LoRA layer into the model
+
+We will now hack the original GPT-2 model and inject LoRA layers into it. Let's
+do a couple of things before doing that:
+
+- Delete previous model;
+- Reset "peak" GPU memory usage using `tf.config.experimental.reset_memory_stats`;
+- Load a new GPT-2 model.
+"""
+
+del gpt2_lm
+del optimizer
+del loss
+
+# This resets "peak" memory usage to "current" memory usage.
+tf.config.experimental.reset_memory_stats("GPU:0")
+
+# Load the original model.
+preprocessor = keras_nlp.models.GPT2CausalLMPreprocessor.from_preset(
+    "gpt2_base_en",
+    sequence_length=128,
+)
+lora_model = keras_nlp.models.GPT2CausalLM.from_preset(
+    "gpt2_base_en",
+    preprocessor=preprocessor,
+)
+
+"""
+We will now override the original query/value projection matrices with our
+new LoRA layers.
+"""
+
+for layer_idx in range(lora_model.backbone.num_layers):
+    # Change query dense layer.
+    decoder_layer = lora_model.backbone.get_layer(f"transformer_layer_{layer_idx}")
+    self_attention_layer = decoder_layer._self_attention_layer
+
+    # Change query dense layer.
+    self_attention_layer._query_dense = LoraLayer(
+        self_attention_layer._query_dense,
+        rank=RANK,
+        alpha=ALPHA,
+        layer_idx=layer_idx,
+        trainable=True,
+    )
+
+    # Change value dense layer.
+    self_attention_layer._value_dense = LoraLayer(
+        self_attention_layer._value_dense,
+        rank=RANK,
+        alpha=ALPHA,
+        layer_idx=layer_idx,
+        trainable=True,
+    )
+
+"""
+Let's now do a forward pass to make sure we still have a valid chain of
+computation.
+"""
+
+lora_model(preprocessor(["they are going to ban LoRA in EU, lol"])[0])
+pass
+
+"""
+Freeze the entire LLM, only the LoRA layers should be trainable.
+"""
+
+for l in lora_model._flatten_layers():
+    lst_of_layers = list(l._flatten_layers())
+
+    if len(lst_of_layers) == 1:  # "leaves of the model"
+        if "A_" in l.name or "B_" in l.name:
+            l.trainable = True
+        else:
+            l.trainable = False
+
+"""
+Print the model's summary and see if the number of non-trainable parameters and
+total parameters are correct.
+
+In a previous section, we had calculated the number of parameters associated with
+the LoRA layers to be 6,144. The total trainable parameters in the model should
+be `num_layers * (query, value) * 6,144 = 12 * 2 * 6,144 = 147,456`. The
+number of non-trainable parameters should be the same as the total number of
+parameters in the original GPT-2 model, which is `124,439,808`.
+"""
+
+lora_model.summary()
+
+"""
+### Fine-tune LoRA GPT-2
+
+Now that we have hacked and verified the LoRA GPT-2 model, let's train it!
+"""
+
+gpu_memory_callback = GPUMemoryCallback(
+    target_batches=[5, 10, 25, 50, 100, 150, 200, 300, 400, 500],
+    print_stats=True,
+)
+
+optimizer = keras.optimizers.AdamW(
```

Comment:
Can we just re-use the original optimizer? Or will that leak some memory from the first training run?

Reply:
Pulled all of it in a function